### ISSUE TITLE
Impossible to compile with AndroidX and Jetifier with minifyEnabled and proguard disabled

### ISSUE THREAD

**Issue by MathieuCalba** [AzureAD/microsoft-authentication-library-for-android#354]
Hello, I have tried to migrate my project to AndroidX and use the jetifier to transform dependencies to use AndroidX. Unfortunately I have an error with the microsoft-authentication-library-for-android library:
```
Execution failed for task ':app:transformClassesWithAndroidGradleClassShrinkerForDebug'.
> SHA-256 digest error for com/microsoft/identity/client/AadAuthority.class
```

I have the issue only in debug builds which have the `minifyEnabled true` & `useProguard false`. And I don’t have the issue on the release builds with `shrinkResources true` & `minifyEnabled true` & `useProguard true`. (I use `minifyEnabled true` on debug builds in my project to remove unused methods to go below the 65K method limit).

I build the debug builds with `./gradlew clean :app:assembleDebug` and the release ones with `./gradlew clean :app:assembleRelease`.

I have created a project reproducing the error:
[AzureADtestAndroidX.zip](https://github.com/AzureAD/microsoft-authentication-library-for-android/files/2375868/AzureADtestAndroidX.zip)

Do you have any idea on what the issue might be?

Thanks

**Comment by iambmelt** [comment]
Hi @MathieuCalba -- Do you have instant run enabled? If yes, can you try disabling it and report your findings?

**Comment by MathieuCalba** [comment]
@iambmelt Instant Run is disabled. I’ve tried to turn it on and off again to see if it changes anything. But no, I still have the same error in Android Studio or from the CLI.

**Comment by iambmelt** [comment]
@MathieuCalba If you find an acceptable solution or workaround, can you update this thread?

If/when we investigate this issue further, I'm happy to update our wiki with guidance if the solution is universally applicable

**Comment by MathieuCalba** [comment]
@iambmelt I opened a bug on the Google issue tracker https://issuetracker.google.com/issues/115556774, and they are looking into it.

**Comment by modislaszlox2** [comment]
I'm having the same issue.

**Comment by modislaszlox2** [comment]
Still not fix for this? How much time do you guys need to fix such a critical issue.

**Comment by jcoronadocc** [comment]
I try  'com.microsoft.aad:adal:1.16.+' , still not working.

**Comment by thomius** [comment]
Just if it makes things clearer, repro steps with a fresh Android Studio installation:

 - Create a new Android Studio project
 - Ensure Instant Run is enabled (File > Settings > Build, Execution, Deployment > Instant Run)
 - Setup MSAL: https://github.com/AzureAD/microsoft-authentication-library-for-android
 - Debug. Things should work out fine.
 - Add the following to gradle.properties:
 - android.useAndroidX=true
 - android.enableJetifier=true
 - Refactor > Migrate to AndroidX
 - Try to start debugging
 - Now compiler reports error "java.lang.SecurityException: SHA-256 digest error for com/microsoft/identity/client/AuthenticationCallback.class"
 - Disable Instant Run
 - Debug
 - Now things work out fine.

**Comment by iambmelt** [comment]
Thanks for the concise repro steps, @thomius -- we'll pick this up once we've triaged and assigned priority on our end.

**Comment by iambmelt** [comment]
Tagging 'Help Wanted' as external contributions to resolve this issue are welcome.

**Comment by kishorsinghgour** [comment]
After migration to AndroidX able to compile code successfully by disabling Instant run but getting _RunTimeException_  

> java.lang.RuntimeException: Unable to stop activity {com.activescaler.imride/com.microsoft.identity.client.AuthenticationActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean com.microsoft.identity.client.AuthenticationActivity$MsalCustomTabsServiceConnection.getCustomTabsServiceIsBound()' on a null object reference
        at android.app.ActivityThread.performStopActivityInner(ActivityThread.java:3890)
        at android.app.ActivityThread.handleStopActivity(ActivityThread.java:3939)
        at android.app.ActivityThread.-wrap25(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1524)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6165)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:888)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:778)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean com.microsoft.identity.client.AuthenticationActivity$MsalCustomTabsServiceConnection.getCustomTabsServiceIsBound()' on a null object reference
        at com.microsoft.identity.client.AuthenticationActivity.onStop(AuthenticationActivity.java:114)
        at android.app.Instrumentation.callActivityOnStop(Instrumentation.java:1311)
        at android.app.Activity.performStop(Activity.java:6862)

when redirected to sign-in/up screen.

**Comment by JohnMAustin78** [comment]
I love how Brian asks such pointed and probing questions that get to the
heart of the matter...

By way of transparency, Brian used to be my teammate at Microsoft's Patsol
team until I went and joined Jake Wharton at Square  :-)

**Comment by JohnMAustin78** [comment]
Gosh darn it Brian, you're my hero... next to Jake of course.

**Comment by MoBo** [comment]
Reminds me of this bug https://github.com/AzureAD/azure-activedirectory-library-for-android/issues/724

**Comment by iambmelt** [comment]
Google tracking issue:
https://issuetracker.google.com/issues/119183822

Proposed workaround PR:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/417

**Comment by joshfriend** [comment]
Now that #417 was merged, is it possible to have a bugfix release? Anyone currently using androidx (like myself) is currently blocked on trying to use this library.

**Comment by joshfriend** [comment]
After much effort, I was able to use the current version with AndroidX:

1. Download and unzip the [jetifier standalone cli](https://developer.android.com/studio/command-line/jetifier#install_jetifier)
1. Download the [msal aar file](https://bintray.com/bintray/jcenter/download_file?file_path=com%2Fmicrosoft%2Fidentity%2Fclient%2Fmsal%2F0.2.2%2Fmsal-0.2.2.aar)
1. Strip the signatures from the aar and jetify it: `./bin/jetifier-standalone -stripSignatures -i msal-0.2.2.aar -o msal-0.2.2.aar`
1. Repeat this process for the [`common-0.0.8`](https://bintray.com/bintray/jcenter/download_file?file_path=com%2Fmicrosoft%2Fidentity%2Fcommon%2F0.0.8%2Fcommon-0.0.8.aar) aar as well
1. In Android Studio, select File > New > New Module and choose "Import .JAR/.AAR package"
1. Select the stripped and jetified `msal-0.2.2.aar` 
1. Also import `common-0.0.8.aar` the same way
1. In the app's `build.gradle`, include msal, common **and all of their child dependencies**:

```groovy
dependencies {
    implementation project(':msal')
    //region MSAL dependencies
    // Have to include these manually because the msal AAR has no way to convey dependencies
    implementation project(':msidentity-common')
    implementation "androidx.browser:browser:1.0.0"
    implementation "com.google.code.gson:gson:2.8.4"
    implementation 'com.nimbusds:nimbus-jose-jwt:5.7'
    //endregion
}
```

**Comment by iambmelt** [comment]
Thanks @joshfriend 

For anyone who'd like to partially automate this process (steps 1-4), the following script can be used

```sh
# First, download the jetifier zip
curl https://dl.google.com/dl/android/studio/jetifier-zips/1.0.0-beta04/jetifier-standalone.zip > jetifier-standalone.zip

# Unzip the file
unzip jetifier-standalone.zip

# Delete it since we're done
rm -fv jetifier-standalone.zip

# Create directories to hold the MSAL binaries + deps
mkdir -p bin/signed/msal bin/signed/common bin/unsigned/msal bin/unsigned/common

# Download the signed binaries to the target dir
curl -Lv --ssl-reqd https://bintray.com/bintray/jcenter/download_file?file_path=com/microsoft/identity/client/msal/0.2.2/msal-0.2.2.aar > bin/signed/msal/msal-0.2.2.aar
curl -Lv --ssl-reqd https://bintray.com/bintray/jcenter/download_file?file_path=com/microsoft/identity/common/0.0.8/common-0.0.8.aar > bin/signed/common/common-0.0.8.aar

# Strip signatures and copy to output dir
./jetifier-standalone/bin/jetifier-standalone -l verbose -stripSignatures -i bin/signed/msal/msal-0.2.2.aar -o bin/unsigned/msal/msal-0.2.2.aar
./jetifier-standalone/bin/jetifier-standalone -l verbose -stripSignatures -i bin/signed/common/common-0.0.8.aar -o bin/unsigned/common/common-0.0.8.aar

# Delete jetifier since we're done
rm -rfv jetifier-standalone
```

And will yield the following output. The binaries in `/bin/unsigned/*` are the ones you'll want to import.

```tree
bin
    ├───signed
    │   ├───common
    │   │       common-0.0.8.aar
    │   │
    │   └───msal
    │           msal-0.2.2.aar
    │
    └───unsigned
        ├───common
        │       common-0.0.8.aar
        │
        └───msal
                msal-0.2.2.aar
```

**Comment by g-sahu** [comment]
Hi, 

Has this issue been fixed? I'm getting a blank page upon signing-in with Microsoft. I'm using MSAL version 0.1.0.

**Comment by iambmelt** [comment]
@g-sahu That doesn't sound like the same issue; others have reported issues with compilation whereas yours sounds like a runtime issue. Can you share why you think it's the same issue?

As for this issue (#354), a preview [workaround](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/417) is available in [`0.3.1-alpha`](https://github.com/AzureAD/microsoft-authentication-library-for-android/releases/tag/v0.3.1-alpha). The actual cause of the issue is jetifier; there are two issues [one is fixed and is in RC](https://issuetracker.google.com/issues/115556774), the [other's status is assigned](https://issuetracker.google.com/issues/119183822)

**Comment by thomius** [comment]
I happened to try this out with the latest Android Studio and everything seemed to work fine both with 0.2.2 and 0.3.1-alpha.

**Comment by iambmelt** [comment]
Thanks for confirming @thomius - I will close this issue for now. Feel free to comment or post a follow-up issue if a regression occurs.

**Comment by g-sahu** [comment]
@iambmelt I was also getting this issue initially while building my project and it got resolved after I disabled instant run in Android Studio. But now, upon running the application I'm getting a blank page after starting the sign-in process with MSAL.
